### ISSUE TITLE
Plugin Architecture (enable & disable, sandboxing)

### ISSUE THREAD

**Issue by MiniGod** [ManiaJS/ManiaJS#122]
When reading this https://github.com/ManiaJS/plugins/issues/15
I thought it would be nice to have general way to disable and enable small plugins like that one, so that they don't have to be bloated with some enable & disable mechanism.
Thoughts?

**Comment by tomvlk** [comment]
Yep, this sort of stuff will be made with #70, only needs some commands to use it without gamemode/script restriction.

It will basically stop the whole plugin, destroy from memory. ^^

**Comment by tomvlk** [comment]
@ManiaJS/core 

Ok, So enable/disable is implemented in the manager.

But there is one problem which we need to think about and, please drop the ideas here.

When disabling the plugin, i'm currently moving it to another array, just sitting there and let all the UI elements being destroyed. But this will not stop the plugin, the callbacks and subscriptions on events will still be called because that's controller from the core and the eventemitters without knowing if the callback is in the disabled plugin.

For me, I see two options.
1. Easy, let the plugin vanish, really remove from memory, no trace of it anymore.
   This requires to reload everytrhing when going to enable it again, which would be good in some situations but tricky in some others.
2. More work, complicated. Let each plugin get it's own eventemitter for every app.\* interface and let it be injected in the plugin. This will make even more variables in the plugins context, which can cause collisions. Also will be bit more expensive for resources.

3 .. ?

**Comment by tomvlk** [comment]
Note. The event emitters will still hold the reference to the function. So nuking is not really an option, it won't nuke the callback function itself. 

Currently I see one option. Create own our event emitter with context separated calls so we can nuke those or at least stop them from executing.

**Comment by tomvlk** [comment]
# Propose new architecture

![maniajs-architecture](https://cloud.githubusercontent.com/assets/3877688/16870915/5e259b70-4a84-11e6-88b4-45dc41e9762d.png)

## Sandbox around plugins(modules)

The sandbox is the way to communicate from the plugin to the core and components. Also for event subscription, the sandbox is used. The sandbox will stop sending those events to the plugin when the plugin is disabled or unloaded. Sandbox will act as an event proxy and facade to core/components.

## Plugin context

The App (core) itself won't be available to the plugin, and the plugin will not have any properties injected anymore, this will prevent potentially property collision with future updates.

The plugin constructor will have the instance of the sandbox available, and need to register itself with the sandbox (give details such as name, paths, versions, dependencies and more stuff.

## Event Emitter and event names

We need to create a clear and good syntax for the events that will be sent. I propose to use it like this:

```
'component/event.name'
```

Where the component is the component name such as server and the eventname is the dynamic event name specific for the component.

Example:

```
'server/player.connect'
```

For filterable and UI events (with dynamic content in the eventname that could collision) we can use a escape pattern like this:

```
'component/event.name[escapedcontent]'
```

Where the escapedcontent can contain `/` and `.` without any problem with resolving.

## Sending queries to server

The current style of sending custom queries is a bit weird, we could better make it like this (from plugin context).

``` js
this.sandbox.server.send('QueryName', ['params']) => Promise
this.sandbox.server.send([
    {query: 'query1', params: ['params']},
    {query: 'query2', params: ['params']}
]) => Promise
this.sandbox.server.chat(options: any) => Promise
this.sandbox.server.query() // Is alias for .send()
```

Any suggestions are welcome.
